docs/05-rfcs/2023-01-11-sdk-spec.md

```diff
@@ -0,0 +1,1070 @@
+   */
+  url: str;
+}
+```
+
+Future extensions: domain and certificate props? support for edge functions?
+
+## Api
+
+The Api resource represents an API Gateway that can be used to manage HTTP routes and run functions in response to requests.
+
+```ts
+struct ApiProps {
+  cors: ApiCorsProps?;
+}
+
+resource Api {
+  init(props: ApiProps = {});
+
+  /**
+   * The api's base url.
+   */
+  url: str;
+
+  /**
+   * Run an inflight whenever a GET request is made to the specified route.
+   */
+  get(route: str, fn: inflight (req: ApiRequest) => ApiResponse, opts: cloud.ApiOnGetProps?): void;
+
+  /**
+   * Run an inflight whenever a POST request is made to the specified route.
+   */
+  post(route: str, fn: inflight (req: ApiRequest) => ApiResponse, opts: cloud.ApiOnPostProps?): void;
+
+  /**
+   * Run an inflight whenever a PUT request is made to the specified route.
+   */
+  put(route: str, fn: inflight (req: ApiRequest) => ApiResponse, opts: cloud.ApiOnPutProps?): void;
+
+  /**
+   * Run an inflight whenever a DELETE request is made to the specified route.
+   */
+  delete(route: str, fn: inflight (req: ApiRequest) => ApiResponse, opts: cloud.ApiOnDeleteProps?): void;
+
+  /**
+   * Run an inflight whenever a PATCH request is made to the specified route.
+   */
+  patch(route: str, fn: inflight (req: ApiRequest) => ApiResponse, opts: cloud.ApiOnPatchProps?): void;
+
+  /**
+   * Run an inflight whenever any request is made to the specified route.
+   */
+  add_route(route: str, fn: inflight (req: ApiRequest) => ApiResponse, opts: cloud.ApiOnRequestProps?): void;
+
+  /**
+   * Make a request to the specified route. Throws if the route hasn't been
+   * defined.
+   */
+  inflight request(route: str, method: HttpMethod, body: Serializable): Serializable;
+}
+
+struct ApiOnGetProps { /* elided */ }
+struct ApiOnPostProps { /* elided */ }
+struct ApiOnPutProps { /* elided */ }
+struct ApiOnDeleteProps { /* elided */ }
+struct ApiOnPatchProps { /* elided */ }
+struct ApiOnRequestProps { /* elided */ }
+
+struct ApiCorsProps {
+  /**
+   * The list of allowed origins.
+   * @example ["https://example.com"]
+   */
+  origins: Array<str>;
+
+  /**
+   * The list of allowed methods.
+   * @example [HttpMethod.GET, HttpMethod.POST]
+   */
+  methods: Array<HttpMethod>;
+
+  /**
+   * The list of allowed headers.
+   * @example ["Content-Type"]
+   */
+  headers: Array<str>;
+
+  /**
+   * The list of exposed headers.
+   * @example ["Content-Type"]
+   */
+  exposed_headers: Array<str>;
+
+  /**
+   * Whether to allow credentials.
+   */
+  allow_credentials: bool;
+}
+
+struct ApiRequest {
+  /** The request's HTTP method. */
+  method: HttpMethod;
+  /** The request's path. */
+  path: str;
+  /** The request's query string. */
+  query: str;
+  /** The path variables. */
+  vars: Map<str, str>;
+  /** The request's body. */
+  body: Serializable;
+  /** The request's headers. */
+  headers: Map<str, str>;
+}
+
+struct ApiResponse {
```

Comment:
Do we need some api here for reading the response iteratively (in case of huge response)?

Reply:
Do you mean if the user wants to create an iterable API? That's a pretty good question... I think the solution here would be for the user to implement pagination in their API. Definitely something we can try to make easier for users, I'll add it as a future extension idea